File: C++/637. Average of Levels in Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
// Time: O(n)
class Solution {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        vector<double> ans;
        queue<TreeNode*> q;
        q.push(root);
        while(!q.empty())
        {
            double sum = 0;
            int cnt = q.size();
            for(int i=0;i<cnt;i++)
            {
                TreeNode* node = q.front();
                sum += node->val;
                q.pop();
                if(node->left!=nullptr)
                    q.push(node->left);
                if(node->right!=nullptr)
                    q.push(node->right);
            }
            ans.push_back(sum/cnt);
        }
        return ans;
    }
};

File: Python/637. Average of Levels in Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
# Time: O(n)
# BFS
class Solution1:
    def averageOfLevels(self, root):
        """
        :type root: TreeNode
        :rtype: List[float]
        """
        ans = list()
        q = list()
        q.append(root)
        while len(q) > 0:
            sum = 0.0
            cnt = len(q)
            for i in range(cnt):
                node = q.pop(0)
                sum += node.val
                if node.left is not None:
                    q.append(node.left)
                if node.right is not None:
                    q.append(node.right)
            ans.append(sum/cnt)
        return ans

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
# DFS
class Solution2:
    def averageOfLevels(self, root):
        """
        :type root: TreeNode
        :rtype: List[float]
        """
        res = list()
        ans = list()

        def count(node, depth):
            if node is None:
                return
            if len(res) <= depth:
                res.append([node.val, 1])
            else:
                res[depth][0] += node.val
                res[depth][1] += 1
            count(node.left, depth+1)
            count(node.right, depth+1)
        count(root, 0)
        for v, d in res:
            ans.append(v/d)
        return ans